packages/sdk/src/cosmwasmclient.ts

```diff
@@ -0,0 +1,228 @@
+      denom: "ucosm",
+    },
+  ],
+  gas: "200000", // 200k
+};
+
+export interface SigningCallback {
+  (signBytes: Uint8Array): Promise<StdSignature>;
+}
+
+interface SigningData {
+  readonly senderAddress: string;
+  readonly signCallback: SigningCallback;
+}
+
+export interface GetNonceResult {
+  readonly accountNumber: number;
+  readonly sequence: number;
+}
+
+export interface PostTxResult {
+  readonly logs: readonly Log[];
+  readonly rawLog: string;
+  readonly transactionHash: string;
+}
+
+export class CosmWasmClient {
+  public static makeReadOnly(url: string): CosmWasmClient {
+    return new CosmWasmClient(url);
+  }
+
+  public static makeWritable(
+    url: string,
+    senderAddress: string,
+    signCallback: SigningCallback,
+  ): CosmWasmClient {
+    return new CosmWasmClient(url, {
+      senderAddress: senderAddress,
+      signCallback: signCallback,
+    });
+  }
+
+  private readonly restClient: RestClient;
+  private readonly signingData: SigningData | undefined;
+
+  private get senderAddress(): string {
+    if (!this.signingData) throw new Error("Signing data not set in this client");
+    return this.signingData.senderAddress;
+  }
+
+  private get signCallback(): SigningCallback {
+    if (!this.signingData) throw new Error("Signing data not set in this client");
+    return this.signingData.signCallback;
+  }
+
+  private constructor(url: string, signingData?: SigningData) {
+    this.restClient = new RestClient(url);
+    this.signingData = signingData;
+  }
+
+  public async chainId(): Promise<string> {
+    const response = await this.restClient.nodeInfo();
+    return response.node_info.network;
+  }
+
+  /**
+   * Returns account number and sequence.
+   *
+   * @param address returns data for this address. When unset, the client's sender adddress is used.
+   */
+  public async getNonce(address?: string): Promise<GetNonceResult> {
+    const account = (await this.restClient.authAccounts(address || this.senderAddress)).result.value;
+    return {
+      accountNumber: account.account_number,
+      sequence: account.sequence,
+    };
+  }
+
+  public async postTx(tx: Uint8Array): Promise<PostTxResult> {
+    const result = await this.restClient.postTx(tx);
+    if (result.code) {
+      throw new Error(`Error when posting tx. Code: ${result.code}; Raw log: ${result.raw_log}`);
+    }
+    return {
+      logs: parseLogs(result.logs) || [],
+      rawLog: result.raw_log || "",
+      transactionHash: result.txhash,
+    };
+  }
+
+  /** Uploads code and returns a code ID */
+  public async upload(wasmCode: Uint8Array, memo = ""): Promise<number> {
+    const storeCodeMsg: MsgStoreCode = {
+      type: "wasm/store-code",
+      value: {
+        sender: this.senderAddress,
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        wasm_byte_code: Encoding.toBase64(wasmCode),
+        source: "",
+        builder: "",
+      },
+    };
+    const fee = defaultUploadFee;
```

Comment:
This is a very nice, simple interface.
I would (in another PR) make the fees configurable in the constructor, like an Options argument, and use default for any that aren't set. So we get the same ease of use, but allow a one-time configuration in the constructor